libdrgn/python/program.c

```diff
@@ -481,6 +481,89 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
 	return drgn_object_copy(ret, &((DrgnObject *)obj)->obj);
 }
 
+static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
+					    struct drgn_module *module,
+					    enum drgn_find_symbol_flags flags,
+					    void *data, struct drgn_symbol_result_builder *builder)
+{
+	struct drgn_error *err = NULL;
+	PyGILState_STATE gstate;
+	_cleanup_pydecref_ PyObject *name_obj = NULL;
+	_cleanup_pydecref_ PyObject *address_obj = NULL;
+	_cleanup_pydecref_ PyObject *one_obj = NULL;
+	_cleanup_pydecref_ PyObject *obj = NULL;
+
+	gstate = PyGILState_Ensure();
+
+	if (flags & DRGN_FIND_SYM_NAME) {
+		name_obj = PyUnicode_FromString(name);
+		if (!name_obj) {
+			err = drgn_error_from_python();
+			goto out;
+		}
+	} else {
+		name_obj = Py_None;
+		Py_INCREF(name_obj);
+	}
+
+	if (flags & DRGN_FIND_SYM_ADDR) {
+		address_obj = PyLong_FromUnsignedLong(addr);
+		if (!address_obj) {
+			err = drgn_error_from_python();
+			goto out;
+		}
+	} else {
+		address_obj = Py_None;
+		Py_INCREF(address_obj);
+	}
+
+	one_obj = PyBool_FromLong(flags & DRGN_FIND_SYM_ONE);
+
+	obj = PyObject_CallFunction(data, "OOO", name_obj, address_obj, one_obj);
+	if (!obj) {
+		err = drgn_error_from_python();
+		goto out;
+	}
+	if (!PyList_Check(obj)) {
+		PyErr_SetString(PyExc_TypeError,
+				"symbol find callback must return list");
+		err = drgn_error_from_python();
+		goto out;
+	}
+
+	size_t len = PyList_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		PyErr_SetString(PyExc_ValueError,
+				"symbol find callback returned multiple elements, but one was requested");
+		err = drgn_error_from_python();
+		goto out;
+	}
+
+	for (size_t i = 0; i < len; i++) {
+		PyObject *item = PyList_GET_ITEM(obj, i);
+		if (!PyObject_TypeCheck(item, &Symbol_type)) {
+			PyErr_SetString(PyExc_ValueError,
+					"symbol find callback elements must be type Symbol");
+			err = drgn_error_from_python();
+			goto out;
+		}
+		_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
+		if (!sym) {
+			err = &drgn_enomem;
+			goto out;
+		}
+		err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
+		if (err)
+			goto out;
+
+		err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+	}
+
+out:
+	PyGILState_Release(gstate);
+	return err;
+}
```

Comment:
This is embarassing, there's a really obvious and really bad memory leak here. We own the reference to the return value of the Python symbol finder function. I need to decref `obj` here in order to free all the symbols and the containing list -- regardless of the success or failure of the function.

Reply:
Oh wait, wow, never mind, this is just illustrating the exact point you made above about putting the `_cleanup_pydecref_` declaration closest to where the object is used...